kernel/src/process.rs

```diff
@@ -351,11 +351,14 @@ pub trait ProcessType {
 
     unsafe fn free(&self, _: *mut u8);
 
-    /// Get a pointer to the grant pointer for this grant number.
+    /// Get the grant pointer for this grant number.
     ///
     /// This will return `None` if the process is inactive and the grant region
     /// cannot be used.
-    unsafe fn grant_ptr(&self, grant_num: usize) -> Option<*mut *mut u8>;
+    ///
+    /// Caution: The grant may not have been allocated yet, so it is possible
+    /// for this grant pointer to be null.
+    fn grant_ptr(&self, grant_num: usize) -> Option<&mut *mut u8>;
```

Comment:
This function allows the caller to get a null pointer, and also returns a mutable reference to change memory allocated to the process. I don't think `unsafe` should be removed, as this seems clearly memory unsafe if misused.

Reply:
- Returning a null pointer is not `unsafe` in Rust terms. What is unsafe is to dereference such pointer or to convert it to a reference, but Rust requires an `unsafe` block for that.
- The reference to a pointer also points to a valid memory location.
- However, what is definitely unsafe is to take an immutable `&self` and returning a mutable `&mut` reference from it. Because one can call:
  ```rust
  let a = process.grant_ptr(0).unwrap();
  let b = process.grant_ptr(0).unwrap();
  // Both `a` and `b` are now mutable references to the same memory location.
  // This is undefined behavior in Rust!
  ```

This could be solved by requiring a `&mut self` parameter. But given that we don't seem to be able to use `&mut self` anytime soon in many of Tock's data structures (this would mean a lot of complex refactoring in Tock), the safe alternative I see is to have two functions.

```rust
fn grant_ptr(&self, grant_num: usize) -> Option<*mut u8>;
fn set_grant_ptr(&self, grant_num: usize, ptr: *mut u8);
```

Reply:
I can to some extent see why that version of `grant_ptr` isn't marked unsafe, but `set_grant_ptr` is still a very unsafe operation, unless there is some way for the implementation to check that the pointer it receives is valid.

It seems like we should encapsulate the low-level pointer management to the Process struct implementation, and instead return the Grant types directly. But this is a little tricky since the grant type depends on the code path used to access the grant.

Reply:
Implemented as suggested in bc15a5ad10